### DISCUSSION TITLE
Speedometer and family

### DISCUSSION THREAD

**Discussion by ullix** [pyqtgraph/pyqtgraph#2050] [Ideas]
I have a matplotlib problem and on googling stumbled over pyqtgraph. Reading though the docs, I see some exciting opportunities for me. However, in the moment I need something I haven't seen in the docs, neither in the examples.

I need a speedometer like widget / graph, like a classic round instrument with rotating pointer. A linear one would also do. In matplotlib I assemble one by combining wedges and an arrow. In pg I have seen arrows, but nothing like wedges. Not even circles, pies, ovals, rectangles. 

Have I missed it or how does one make /use /construct those graph primitives in pg?

**Comment by j9ac9k** [comment]
Hi @ullix

PyQtGraph certainly does not offer a spedo-meter graphics item out of the box as you've noticed.  The primitive shapes you describe don't actually come from pyqtgraph but from Qt itself, specifically from the QPainter draw methods: https://doc.qt.io/qt-5/qpainter.html  for a higher level description, check out their QGraphicsView page: https://doc.qt.io/qt-5/qgraphicsview.html  All this applies to pyqtgraph.  Where pyqtgraph accels is we provide some sane defaults, view areas that scale, and easier ways to pass numpy arrays to plots.

Some of the polygons and rectangle drawing has rules for filling as well.  For another example of a "custom" graph, check out the `CustomGraphItem` example (while it's not the best example to highlight doing text and such, it should still be able to get you started).

Hope this helps!

**Comment by ullix** [comment]
Thanks for the helping info. Yes indeed, when one has done a lot with PyQt(5) - as I have  - there is clear attractiveness of pyqtgraph! Though QT has its own challenges.

I figured out I need to use QPieSeries() with a hole and append slices, add this to a QChart, view this with a QChartView, and then can add this as a widget to a Q-layout. Phew. But it works.

Except that QPieSeries has a "feature", which makes it unusable for this purpose. For a speedometer you need an arc roughly beginning in the West, going through North, and ending in east. But QPieSeries has 0 degrees in North, and cannot draw a pie crossing this 0 degrees point!  When you want to draw a pie spanning 20 degrees from 350 degrees through 10 degrees, then the resulting arc goes through south, and has a span of 340 degrees!

Too bad. Back to matplotlib.

**Reply by j9ac9k** [reply]
You can generally rotate graphics items by applying a `QTransform.rotate(GraphicsItem, angle)` on them; so if the rotation is the only issue, I would start from there.

I suspect for customization and such, you will want to create a new GraphicsItem instance, that uses QPainter.drawShape primatives if you want to have more extensive customization though.

**Reply by ullix** [reply]
I tried the rotation, and yes, it works. But I agree, this approach is not the way to go. It was, however, easy to implement in matplotlib.

**Reply by j9ac9k** [reply]
If you know how to do this with matplotlib, you should probably stick w/ that... this use case doesn't conform that well to PyQtGraph's strengths (fast w/ large datasets, viewbox scaling, mouse interactivity); if you were trying to get it to run on more of an embedded platform or embedded within a Qt application, it might still be worthwhile to attempt depending if the downsides of using matplotlib don't present any deal breakers.  If you do so, I would look at following the custom graphics item example to guide you at your own implementation.

**Reply by sem-geologist** [reply]
I am not sure or would be not so fast to say that gauge widget would not fit the PyQtGraph.
Gauges are quite universal mechanical devices (or software widgets, resembling those mechanical) to show most recently probed or actual state of something. What @ullix wanted is radial gauge for speedometer, but that is only one of many usage cases for gauges, and gauges do not need to be radial, it can be also (progress-)bar-like. Gauges has very huge advantage in that value and its position (or simply the situation of parameter) in given scale can be near instantly assessed visually (low fuel, low battery, full battery, movement (speed) of object/fluid/gas/potential out from/within safe limits, near danger zone, at position for further procedures to kick-in and many many more). I have this impression that PyQtGraph is used mostly for scientific and industrial software, and it is not excluded for instrument control, and with near real-time capabilities for sure for monitoring. PyQtGraph added lately more convenient control features (i.e. expanded parameter tree), it also have some controls which hardly are connected with viewbox'es (i.e. software joystick). The good GUI practice is to give clear feedback when changing some parameter, and gauges are excellent at that. Of course they are useless for offline big-data plotting, but for real time monitoring they have its point of existence. 

There is still quite a lack of well thought gauge library in PyQt/PySide.
In my opinion gauges should be simple to read, use, and setup. Gauges should have possibility to colour zones independently from direction (in some cases low value is danger, in other high value means danger), it should be easy to mark zones (optimal range).

Ultimately it would be nice for monitoring to feed same data to gauge and real-time plotting. In case scrolling back in time, the gauge would still show the most recent value; for understanding of some processes and correct meaning of current value it is important to understand how system progressed to that point. Sometimes there is not much time to do that, and while inspecting the data before that point in real-time plot (scrolling and zooming to previous events), it is easy to loose track of what is going on at the very current moment, which in some cases can be dangerous.
One of such real-life example would be abusing Schottky field emission electron emitter, where in some cases current value, and history of value needs to be seen simultaneously for making the right decision in control. The gauge alongside (or as inset) to real-time plotting viewbox would be ideal solution. Of course for same solution two real-time plots could be used, one updating to last committed measurement, and sliding the view, other for inspection, but that takes more screen space and is harder to comprehend.

That's my two cents.

**Comment by sem-geologist** [comment]
hi @ullix , I dont think QPieSeries and chart is optimal aproach. what you should search is Gauge widget, and Qt has one ready built-in however in QML (`QtQuick.Extras 1.4` if to be more precise) (1); there are plenty c++/Qt gauge widgets licensed with gpl/lgpl (2), to use that with python, bindings should be generated (by yourself); but as You are using pyqtgraph, probably you are looking for python-ready solution. I found one, quite customisable, apache licensed (I had no opportunity to try it):
https://github.com/StefanHol/AnalogGaugeWidgetPyQt

BTW, I was looking for such kind of gauge widgets for some custom UPS monitoring tool.

**Reply by sem-geologist** [reply]
Another free python ready library with Analog Gauge is this:
https://github.com/KhamisiKibet/QT-PyQt-PySide-Custom-Widgets
albeit it is aimed only for PySide2, which Could be a problem.
Beware, an example is over-shiny (for my liking at least), however it is possible to make it more flat and less shiny.

**Reply by ullix** [reply]
Thanks for the links! I took the StefanHol gauge, and adapted it. It is beginning to look like real ;-), though I need log-scale, which is not trivial. But already far ahead from initial Pie approach.

**Comment by ullix** [comment]
Gauges are technical instruments which come in very different incarnation and for very different purposes. Mine is for the display of count rates from a Geiger counter. As those rates may vary from a low of 10 CPM (Counts Per Minute) up to 100000 CPM, a log scale is an appropriate choice.  

I finished my gauge to a satisfactory degree ![image](https://user-images.githubusercontent.com/9335726/139697998-2b8a7014-1eb9-41a6-b563-61cbb413a622.png), and in the end it is basically a QPainter job with some other Qt sprinkles. 

So, it took me a detour in getting there - thanks for your guidance.

Whether a gauge is something that pg should offer may be debatable. But I am leaning towards @sem-geologist 's perspective. Given that you claim in your docu's first intro-paragraph: "... and 2) to provide tools to aid in rapid application development " it would seem to me that such tools as a gauge and many others would also be needed for rapid application development!

For me the whole thing was valuable despite or perhaps because of the detour. I experimented a bit with pg for plotting data, and, yes, easy enough to start and to get good results. I haven't dug into customization. Mostly because my most relevant graph requires 2 Y-axis with different units, which is, so my understanding, not (yet) possible with pg.

I'll keep an eye on the site!

**Reply by j9ac9k** [reply]
I have a mechanical engineering background; and once upon a time partook in motorsports, which I say as to illustrate I understand the importance of this kind of data visualization, and I am not attempting to suggest they wouldn't be useful.

That said, there are a few problematic entries with integrating this into pyqtgraph directly.

* It does not conform to the areas pyqtgraph accelerates in
  * Does not go into a scaled plot area
  * Plot elements would not have a high degree of mouse interactivity (maybe hover events?)
  * Does not take large arrays of data
* There are existing solutions that appear maintained and have high degrees of customization that pyqtgraph is unlikely to be equivalent with (although if there is something that can be changed in PyQtGraph to allow for embedding of these kinds of graphics items, I'd like to hear that).
  
To get the level of customization a user would want, they would likely need to work w/ the underlying Qt framework methods a fair amount as it is, so there is really not much PyQtGraph could do here.  Or another way of saying this, if I was to attempt to implement something like this in PyQtGraph, I'm not sure it would involve importing another part of PyQtGraph at all; _maybe_ [GraphicsObject](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/GraphicsObject.py)

Another way of thinking about is that pyqtgrah doesn't really offer the ability to draw primitive geometries, we rely on Qt to do that, and that is what end-users would likely need to do anyway to get the criteria they want.

That said, I think this would be a good use case to integrate as another "custom graphics item" example, just like our [candlestick plot](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/customGraphicsItem.py)

Other maintainers may certainly feel differently than I do; but in terms of upcoming plots to add, I have my sights set on what pandas would need for a plotting backend, which would require a variety of plots not currently in the library.

**Comment by ullix** [comment]
I agree with everything you said, except your second bullet point: I don't see many existing solutions. But whatever there is, if you could make some place available for them to advertise - in your docu at least - that would be great!

I took a peek at the candlestick code and saw `from pyqtgraph import QtCore, QtGui`.

Goodness, are you having your own versions of PyQt???

**Reply by sem-geologist** [reply]
@ullix yes, it does, but that is not a bad thing, `pyqt` library does not scratch all itches...

@j9ac9k , This is discussion section, I don't request this feature. At least I  want to discuss and see if that is more suitable or less suitable to `PyQtGraph`. Currently I see two people which would be interested in such feature (myself and @ullix ), which is good start.

now if look into those bullet points...
* pyqtgraph offers lots of things which does not go into scaled plot area (color button, colorbar, joystick widget, parameter tree, table, array widgets, ....)
* That's partly true and even if not.
* Actually, it would be nice If it could accept large array and show always the last item in the array (most recently added), it would be wonderful that it would use same array as plotDataItems, so that it would stay in sync.
* Yes there are some limited ready solutions with some or "too-much" customization more in visual (eye-candy... or eye-sore) than functional directions. Some problems are like lack of floating-point support, no log axis, too many confusing visualization options fitted more for car speedometer, shiny web app or some hipster stuff. 

Few more things to consider:
* qwt (similar technical-aimed software library) has radial diagrams and gauges.
* Actually as I think about implementation of it in pyqtgraph - Gauges could reuse some `pyqtgraph` code: i.e. `AxisItem` directly (linear gauges) and partially (radial gauges) which i.e. takes care of log scaling.

As I have some  pressure gathering on me currently to make some GUI monitoring and safety tools for some technical instrumentation in our lab, which should do real-time plotting and it should also have gauges for presentation of most recent data... let's say I am tempted to make more broad functional Gauge implementation, in particularly that I will need different kind of gauges (linear, radial, bar-like...).

I already have some ideas how to abstract functionality of Gauges (so that gauges could be interchangeable).

If I get  no clear "No", that means I could continue this discussion on details addressing types of gauges, its common elements, possible approaches. Maybe there would jump more people with more ideas, which should be taken in consideration before starting writing implementation.

As for `pandas`... good to know. What is current situation of plotting simple x y data with pandas datetime type as index? Some time ago when I was trying I had to cast it into unix timestamps for that to work.

**Reply by sem-geologist** [reply]
Just to add to the above, the functionality partly overlaps also radial plots (x as angle, y as intensity), which are missing in `pyqtgraph`...

**Reply by j9ac9k** [reply]
> I took a peek at the candlestick code and saw `from pyqtgraph import QtCore, QtGui`.
> 
> Goodness, are you having your own versions of PyQt???

Ha, heavens no, we have our own abstraction module, much like QtPy but nowhere near as "complete".  This is what lets our library work with PyQt5, PyQt6, PySide2, PySide6  without having if/elif/elif/else statements throughout.  While pyqtgraph users could certainly use pyqtgraph.Qt as an abstraction layer for their own applications, we do not recommend it since we don't claim, nor attempt to test, that our shims have complete coverage.  We just check against parts of Qt we use ourselves.

**Reply by j9ac9k** [reply]
If a gauge-like plot widget was developed, that made use of other parts of the library such as AxisItem, and/or introduced new plot types that could be abstracted (like radial plots), then the biggest concerns I have would be addressed and I would be fine with incorporating that sort of feature.

> As I have some pressure gathering on me currently to make some GUI monitoring and safety tools for some technical instrumentation in our lab, which should do real-time plotting and it should also have gauges for presentation of most recent data... let's say I am tempted to make more broad functional Gauge implementation, in particularly that I will need different kind of gauges (linear, radial, bar-like...).

> I already have some ideas how to abstract functionality of Gauges (so that gauges could be interchangeable).

So I don't mean to come across as outright rejecting a submission before one would be made, or to say it's not worth our while.  I think incorporating some kind of radial plot functionality, and perhaps having the gauge be an example, would be a great place to strat.

> As for pandas... good to know. What is current situation of plotting simple x y data with pandas datetime type as index? Some time ago when I was trying I had to cast it into unix timestamps for that to work.

I know we have some support for datetime plots, I haven't used it myself much, but our date-time plot require the dates to be the x-axis.  

I also want to point out that I'm not the sole decider here, we have other maintainers who (very likely) have other opinions.  

If the question is "can pyqtgraph add gauge widgets?" ...we're going to be hesitant, but if the question is "can we add radial plots and then add an example of those radial plots that makes a gauge" I think that's a different story.

**Comment by sem-geologist** [comment]
I just checked how `qwt` does all that, so my intuition on reusing AxisItem is rather right direction (albeit in qwt, it reuses from abstract axis part, but on python it could simply subclass `AxisItem`, see https://qwt.sourceforge.io/class_qwt_abstract_scale.html)
